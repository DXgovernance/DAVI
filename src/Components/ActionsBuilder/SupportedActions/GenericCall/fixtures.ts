import {
  ApproveSendTokens,
  DecodedCall,
  SupportedAction,
} from 'Components/ActionsBuilder/types';
import ERC20ABI from 'contracts/ERC20.json';
import { BigNumber, utils } from 'ethers';

const ERC20Contract = new utils.Interface(ERC20ABI.abi);

export const genericDecodedCallMock: DecodedCall = {
  from: '0x0000000000000000000000000000000000000000',
  callType: SupportedAction.GENERIC_CALL,
  function: ERC20Contract.getFunction('increaseAllowance'),
  to: '0x3f943f38b2fbe1ee5daf0516cecfe4e0f8734351',
  value: BigNumber.from(0),
  args: {
    param1: BigNumber.from(10000),
    param2: BigNumber.from(10000),
    param3: '0x3f943f38b2fbe1ee5daf0516cecfe4e0f8734351',
  },
};

export const approvetokenSpendingMock: ApproveSendTokens = {
  amount: BigNumber.from(1000),
  token: '0x3f943f38b2fbe1ee5daf0516cecfe4e0f8734351',
};
