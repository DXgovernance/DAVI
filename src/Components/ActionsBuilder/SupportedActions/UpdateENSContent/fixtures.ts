import { BigNumber, utils } from 'ethers';
import { DecodedCall } from 'Components/ActionsBuilder/types';
import { SupportedAction } from 'Components/ActionsBuilder/types';
import ENSPublicResolver from 'contracts/ENSPublicResolver.json';

const ENSPublicResolverContract = new utils.Interface(ENSPublicResolver.abi);

export const mockDecodedCallUpdateENSContent: DecodedCall = {
  from: '0x0000000000000000000000000000000000000000',
  callType: SupportedAction.ENS_UPDATE_CONTENT,
  function: ENSPublicResolverContract.getFunction('setContenthash'),
  to: '0x3f943f38b2fbe1ee5daf0516cecfe4e0f8734351',
  value: BigNumber.from(0),
  args: {
    node: '0x0000000000000000000000000000000000000000',
    hash: '0x0000000000000000000000000000000000000000',
  },
  optionalProps: {
    ensName: 'test',
    ipfsHash: 'Qm',
  },
};
