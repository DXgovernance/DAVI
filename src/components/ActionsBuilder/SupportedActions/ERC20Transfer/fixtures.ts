import { DecodedCall, SupportedAction } from 'components/ActionsBuilder/types';
import ERC20ABI from 'contracts/ERC20.json';
import { BigNumber, utils } from 'ethers';

const ERC20Contract = new utils.Interface(ERC20ABI.abi);

export const erc20TransferEmptyDecodedCallMock: DecodedCall = {
  from: '0x0000000000000000000000000000000000000000',
  callType: SupportedAction.ERC20_TRANSFER,
  function: ERC20Contract.getFunction('transfer'),
  to: '',
  value: BigNumber.from(0),
  args: {
    _to: '',
    _value: BigNumber.from(0),
  },
};

export const erc20TransferDecodedCallMock: DecodedCall = {
  from: '0x0000000000000000000000000000000000000000',
  callType: SupportedAction.ERC20_TRANSFER,
  function: ERC20Contract.getFunction('transfer'),
  to: '0xfFb1cd0F95368DDd06D556161c5D3d9f0f4Fe6d2',
  value: BigNumber.from(0),
  args: {
    _to: '0xaF8eB8C3A5d9d900AA0B98e3Df0bcC17d3C5F698',
    _value: BigNumber.from('5000000000000000000'),
  },
};
