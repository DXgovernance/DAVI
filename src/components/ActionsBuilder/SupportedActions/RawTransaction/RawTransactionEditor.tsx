import { Button } from 'components/primitives/Button';
import { AddressInput } from 'components/primitives/Forms/AddressInput';
import {
  Control,
  ControlLabel,
  ControlRow,
} from 'components/primitives/Forms/Control';
import { ErrorLabel } from 'components/primitives/Forms/ErrorLabel';
import { TokenAmountInput } from 'components/primitives/Forms/TokenAmountInput';
import { BigNumber } from 'ethers';
import { useMemo } from 'react';
import { Controller, useForm } from 'react-hook-form';
import { useTranslation } from 'react-i18next';
import { ActionEditorProps } from 'components/ActionsBuilder/SupportedActions';
import validateRawTransaction from './validateRawTransaction';
import { RawTransactionValues } from './types';
import { TextArea } from 'components/primitives/Forms/TextArea';

const RawTransactionEditor: React.FC<ActionEditorProps> = ({
  decodedCall,
  onSubmit,
}) => {
  const parsedData = useMemo(() => {
    return {
      to: decodedCall?.to,
      value: decodedCall?.value,
      data: decodedCall?.optionalProps?.data,
    };
  }, [decodedCall]);

  const { t } = useTranslation();

  const { control, handleSubmit } = useForm({
    resolver: validateRawTransaction,
    context: { t },
    defaultValues: {
      to: parsedData?.to,
      value: parsedData?.value,
      data: parsedData?.data,
    },
  });

  const submitAction = (values: RawTransactionValues) => {
    if (!values.data) values.data = '0x00';
    if (!values.value) values.value = BigNumber.from(0);

    onSubmit([
      {
        ...decodedCall,
        to: values.to,
        value: values.value,
        optionalProps: {
          data: values.data,
        },
      },
    ]);
  };

  return (
    <div>
      <form onSubmit={handleSubmit(submitAction, console.error)}>
        <Controller
          name="to"
          control={control}
          render={({ field: { ref, ...field }, fieldState }) => {
            const { error } = fieldState;

            return (
              <Control>
                <ControlLabel>{t('toAddress')}</ControlLabel>
                <ControlRow>
                  <AddressInput
                    {...field}
                    data-testid="input-to"
                    isInvalid={!!error}
                    placeholder={t('ethereumAddress')}
                  />
                </ControlRow>
                {!!error && (
                  <ErrorLabel data-testid="error-to" margin="0.5rem 0">
                    {error.toString()}
                  </ErrorLabel>
                )}
              </Control>
            );
          }}
        />

        <Controller
          name="value"
          control={control}
          render={({ field: { ref, ...field }, fieldState }) => {
            const { error } = fieldState;

            return (
              <Control>
                <ControlLabel>{`${t('value')} (${t('in')} wei)`}</ControlLabel>
                <ControlRow>
                  <TokenAmountInput
                    {...field}
                    data-testid="input-value"
                    decimals={0}
                    isInvalid={!!error}
                  />
                </ControlRow>

                {!!error && (
                  <ErrorLabel data-testid="error-value" margin="0.5rem 0">
                    {error.toString()}
                  </ErrorLabel>
                )}
              </Control>
            );
          }}
        />

        <Controller
          name="data"
          control={control}
          render={({ field: { ref, ...field }, fieldState }) => {
            const { error } = fieldState;

            return (
              <Control>
                <ControlLabel>{`${t('data')} (hex)`}</ControlLabel>
                <ControlRow>
                  <TextArea
                    {...field}
                    data-testid="input-data"
                    spellCheck={false}
                    isInvalid={!!error}
                    placeholder={t('data')}
                  />
                </ControlRow>
                {!!error && (
                  <ErrorLabel data-testid="error-data" margin="0.5rem 0">
                    {error.toString()}
                  </ErrorLabel>
                )}
              </Control>
            );
          }}
        />

        <Button
          m="1rem 0 0"
          fullWidth
          data-testid="submit-rawtransaction"
          type="submit"
        >
          {t('saveAction')}
        </Button>
      </form>
    </div>
  );
};

export default RawTransactionEditor;
