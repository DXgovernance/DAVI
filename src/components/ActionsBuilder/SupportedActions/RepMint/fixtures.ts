import { DecodedCall, SupportedAction } from 'components/ActionsBuilder/types';
import { BigNumber, utils } from 'ethers';
import ERC20SnapshotRep from 'contracts/ERC20SnapshotRep.json';

const ERC20SnapshotRepContract = new utils.Interface(ERC20SnapshotRep.abi);

export const repMintEmptyDecodedCallMock: DecodedCall = {
  from: '0x0000000000000000000000000000000000000000',
  callType: SupportedAction.REP_MINT,
  function: ERC20SnapshotRepContract.getFunction('mint'),
  to: '',
  value: BigNumber.from(0),
  args: {
    to: '',
    amount: BigNumber.from(0),
  },
};

export const repMintDecodedCallMock: DecodedCall = {
  from: '0x0000000000000000000000000000000000000000',
  callType: SupportedAction.REP_MINT,
  function: ERC20SnapshotRepContract.getFunction('mint'),
  to: '0x3f943f38b2fbe1ee5daf0516cecfe4e0f8734351',
  value: BigNumber.from(0),
  args: {
    to: '0x3f943f38b2fbe1ee5daf0516cecfe4e0f8734351',
    amount: BigNumber.from(10000),
  },
};
