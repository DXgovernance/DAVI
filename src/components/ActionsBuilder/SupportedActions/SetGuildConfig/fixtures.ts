import { BigNumber, utils } from 'ethers';
import { SupportedAction } from 'components/ActionsBuilder/types';
import BaseERC20Guild from 'contracts/BaseERC20Guild.json';

export const mockBigNumber = BigNumber.from(100000000);
export const BaseERC20GuildContract = new utils.Interface(BaseERC20Guild.abi);

export const emptyDecodedCall = {
  from: '',
  callType: SupportedAction.SET_GUILD_CONFIG,
  function: BaseERC20GuildContract.getFunction('setConfig'),
  to: '',
  value: BigNumber.from(0),
  args: {
    _proposalTime: '',
    _timeForExecution: '',
    _votingPowerPercentageForProposalExecution: '',
    _votingPowerPercentageForProposalCreation: '',
    _voteGas: '',
    _maxGasPrice: '',
    _maxActiveProposals: '',
    _lockTime: '',
    _minimumMembersForProposalCreation: '',
    _minimumTokensLockedForProposalCreation: '',
  },
  optionalProps: {},
};
