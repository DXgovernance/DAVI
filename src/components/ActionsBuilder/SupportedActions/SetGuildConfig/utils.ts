import { FIELDS } from './constants';
import { BigNumber } from 'ethers';
import { SetGuildConfigFields, ControlField } from './types';

export const bn = (n: string | number | BigNumber) => BigNumber.from(n);

type UpdatedValuesReturn = {
  [key in ControlField]: {
    newValue: BigNumber;
    currentValue: BigNumber;
  };
};
export const getUpdatedValues = (
  current: SetGuildConfigFields,
  modifyed: SetGuildConfigFields
): UpdatedValuesReturn | {} => {
  if (!current || !modifyed) return {};
  return Object.keys(current).reduce((acc, key) => {
    if (!FIELDS.map(f => f.name).includes(key)) return acc;
    const currentValue = bn(current[key] || 0);
    const newValue = bn(modifyed[key] || 0);

    if (!currentValue.eq(newValue)) {
      return {
        ...acc,
        [key]: { newValue, currentValue },
      };
    }
    return acc;
  }, {});
};

export const pickUpdatedOrDefaultValue = (
  current: BigNumber,
  modifyed: BigNumber
): BigNumber => {
  if (BigNumber.isBigNumber(current)) {
    if (BigNumber.isBigNumber(modifyed) && !bn(modifyed).eq(current)) {
      return bn(modifyed);
    }
    return bn(current);
  } else if (BigNumber.isBigNumber(modifyed)) {
    return bn(modifyed);
  }
  return bn(0);
};
