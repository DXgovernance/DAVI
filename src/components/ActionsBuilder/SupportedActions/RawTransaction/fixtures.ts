import { DecodedCall, SupportedAction } from 'components/ActionsBuilder/types';
import { BigNumber } from 'ethers';
import { MOCK_ADDRESS } from 'hooks/Guilds/ens/fixtures';

export const rawDataCallMock: DecodedCall = {
  from: '0x0000000000000000000000000000000000000000',
  callType: SupportedAction.RAW_TRANSACTION,
  function: null,
  to: MOCK_ADDRESS,
  value: BigNumber.from('1000000000000000000'),
  args: null,
  optionalProps: {
    data: '0x01',
  },
};
