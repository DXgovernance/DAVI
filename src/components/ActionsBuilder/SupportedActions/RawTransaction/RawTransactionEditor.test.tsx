import { fireEvent, waitFor } from '@testing-library/react';
import { BigNumber } from 'ethers';
import { MOCK_ADDRESS, MOCK_ENS_NAME } from 'hooks/Guilds/ens/fixtures';
import { act } from 'react-dom/test-utils';
import { render } from 'utils/tests';
import { rawDataCallMock } from './fixtures';
import RawTransactionEditor from './RawTransactionEditor';

jest.mock('wagmi', () => ({
  __esModule: true,
  useEnsResolver: ({ name, chainId }) => {
    if (!name || name === '.eth') {
      return {
        data: {
          name: null,
          address: null,
        },
      };
    } else {
      return {
        data: {
          name: MOCK_ENS_NAME,
          address: MOCK_ADDRESS,
        },
      };
    }
  },
  useNetwork: () => ({
    chain: {
      id: 1,
      blockExplorers: {
        default: {
          url: 'https://etherscan.io',
        },
      },
    },
  }),
  useEnsName: ({ address, chainId }) => {
    if (!address) {
      return {
        data: null,
      };
    } else {
      return {
        data: MOCK_ENS_NAME,
      };
    }
  },
  useEnsAddress: ({ name, chainId }) => {
    if (!name || name === '.eth' || name === 'invalidAddress.eth') {
      return {
        data: null,
      };
    } else {
      return {
        data: MOCK_ADDRESS,
      };
    }
  },
  useContractRead: () => ({
    data: 'e30101701220e09973e8c9e391cb063bd6654356e64e0ceced7858a29a8c01b165e30a5eb5be',
  }),
  useProvider: () => ({
    getNetwork: jest.fn(),
  }),
  chain: {
    mainnet: {},
  },
  useContractReads: () => ({
    data: [{}],
  }),
  chainId: {
    localhost: 1337,
  },
}));

describe('RawTransactionEditor', () => {
  describe('valid inputs', () => {
    it('should match snapshot', () => {
      const { container } = render(
        <RawTransactionEditor
          decodedCall={rawDataCallMock}
          onSubmit={jest.fn()}
        />
      );

      expect(container).toMatchSnapshot();
    });

    it('should submit if all fields are present', async () => {
      const mockOnSubmit = jest.fn();

      const data = { ...rawDataCallMock };
      const { findByTestId } = render(
        <RawTransactionEditor decodedCall={data} onSubmit={mockOnSubmit} />
      );

      await act(async () => {
        const submitButton = await findByTestId('submit-rawtransaction');
        fireEvent.submit(submitButton);
      });

      await waitFor(() => {
        expect(mockOnSubmit).toHaveBeenCalled();
      });
    });

    it('should submit if there is data and value is zero', async () => {
      const mockOnSubmit = jest.fn();

      const data = { ...rawDataCallMock, value: BigNumber.from(0) };
      const { findByTestId } = render(
        <RawTransactionEditor decodedCall={data} onSubmit={mockOnSubmit} />
      );

      const submitButton = await findByTestId('submit-rawtransaction');
      fireEvent.submit(submitButton);

      await waitFor(() => {
        expect(mockOnSubmit).toHaveBeenCalled();
      });
    });

    it('should submit if there is value and no data', async () => {
      const mockOnSubmit = jest.fn();
      const data = {
        ...rawDataCallMock,
        value: BigNumber.from(111),
        optionalProps: { data: '' },
      };
      const { findByTestId } = render(
        <RawTransactionEditor decodedCall={data} onSubmit={mockOnSubmit} />
      );
      const submitButton = await findByTestId('submit-rawtransaction');
      fireEvent.submit(submitButton);
      await waitFor(() => {
        expect(mockOnSubmit).toHaveBeenCalled();
      });
    });

    it('should submit if there is value and data is zero', async () => {
      const mockOnSubmit = jest.fn();
      const data = {
        ...rawDataCallMock,
        value: BigNumber.from(111),
        optionalProps: { data: '0x00' },
      };
      const { findByTestId } = render(
        <RawTransactionEditor decodedCall={data} onSubmit={mockOnSubmit} />
      );
      const submitButton = await findByTestId('submit-rawtransaction');
      fireEvent.submit(submitButton);
      await waitFor(() => {
        expect(mockOnSubmit).toHaveBeenCalled();
      });
    });
  });

  describe('fail test', () => {
    it('should fail if to address is empty', async () => {
      const mockOnSubmit = jest.fn();

      const data = { ...rawDataCallMock, to: '' };
      const { findByTestId } = render(
        <RawTransactionEditor decodedCall={data} onSubmit={mockOnSubmit} />
      );

      await act(async () => {
        const submitButton = await findByTestId('submit-rawtransaction');
        fireEvent.submit(submitButton);
      });

      await waitFor(() => {
        expect(mockOnSubmit).not.toHaveBeenCalled();
      });
    });

    it('should fail if to address is invalid', async () => {
      const mockOnSubmit = jest.fn();

      const data = { ...rawDataCallMock, to: 'invalidAddress' };
      const { findByTestId } = render(
        <RawTransactionEditor decodedCall={data} onSubmit={mockOnSubmit} />
      );

      await act(async () => {
        const submitButton = await findByTestId('submit-rawtransaction');
        fireEvent.submit(submitButton);
      });

      await waitFor(() => {
        expect(mockOnSubmit).not.toHaveBeenCalled();
      });
    });

    it('should fail if to address is empty', async () => {
      const mockOnSubmit = jest.fn();

      const data = { ...rawDataCallMock, to: 'invalidAddress' };
      const { findByTestId } = render(
        <RawTransactionEditor decodedCall={data} onSubmit={mockOnSubmit} />
      );

      await act(async () => {
        const submitButton = await findByTestId('submit-rawtransaction');
        fireEvent.submit(submitButton);
      });

      await waitFor(() => {
        expect(mockOnSubmit).not.toHaveBeenCalled();
      });
    });

    it('should fail if value is zero and data is null', async () => {
      const mockOnSubmit = jest.fn();

      const data = {
        ...rawDataCallMock,
        value: BigNumber.from(0),
        optionalProps: { data: '' },
      };
      const { findByTestId } = render(
        <RawTransactionEditor decodedCall={data} onSubmit={mockOnSubmit} />
      );

      await act(async () => {
        const submitButton = await findByTestId('submit-rawtransaction');
        fireEvent.submit(submitButton);
      });

      await waitFor(() => {
        expect(mockOnSubmit).not.toHaveBeenCalled();
      });
    });

    it('should fail if value and data is zero', async () => {
      const mockOnSubmit = jest.fn();

      const data = {
        ...rawDataCallMock,
        value: BigNumber.from(0),
        optionalProps: { data: '0x0' },
      };
      const { findByTestId } = render(
        <RawTransactionEditor decodedCall={data} onSubmit={mockOnSubmit} />
      );

      await act(async () => {
        const submitButton = await findByTestId('submit-rawtransaction');
        fireEvent.submit(submitButton);
      });

      await waitFor(() => {
        expect(mockOnSubmit).not.toHaveBeenCalled();
      });
    });

    it('should fail if data is invalid', async () => {
      const mockOnSubmit = jest.fn();

      const data = {
        ...rawDataCallMock,
        value: BigNumber.from(0),
        optionalProps: { data: '0xNotValidHex' },
      };
      const { findByTestId } = render(
        <RawTransactionEditor decodedCall={data} onSubmit={mockOnSubmit} />
      );

      await act(async () => {
        const submitButton = await findByTestId('submit-rawtransaction');
        fireEvent.submit(submitButton);
      });

      await waitFor(() => {
        expect(mockOnSubmit).not.toHaveBeenCalled();
      });
    });

    it('should fail if data is of odd length', async () => {
      const mockOnSubmit = jest.fn();

      const data = {
        ...rawDataCallMock,
        value: BigNumber.from(0),
        optionalProps: { data: '0x0' },
      };
      const { findByTestId } = render(
        <RawTransactionEditor decodedCall={data} onSubmit={mockOnSubmit} />
      );

      await act(async () => {
        const submitButton = await findByTestId('submit-rawtransaction');
        fireEvent.submit(submitButton);
      });

      await waitFor(() => {
        expect(mockOnSubmit).not.toHaveBeenCalled();
      });
    });
  });
});
