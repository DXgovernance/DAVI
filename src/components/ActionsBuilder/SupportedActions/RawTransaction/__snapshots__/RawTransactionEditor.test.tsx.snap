// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RawTransactionEditor valid inputs should match snapshot 1`] = `
.c12 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: center;
  cursor: pointer;
  font-size: 14px;
  border: 1px solid #303338;
  background-color: #295FF4;
  color: #fff;
  border-radius: 32px;
  padding: 0.5rem 0.8rem;
  margin: 0.2rem;
  width: 100%;
  margin: 1rem 0 0;
}

.c12:disabled {
  color: initial;
  opacity: 0.4;
  cursor: auto;
}

.c12:hover:enabled {
  border-color: #fff;
}

.c12:active:enabled {
  border: 1px solid #303338;
}

.c6 {
  height: 1.6rem;
  width: 1.6rem;
}

.c8 {
  border: none;
  background: none;
  background-color: transparent;
  box-shadow: none;
  width: 100%;
  box-sizing: border-box;
  outline: none;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  box-sizing: border-box;
  width: 100%;
  border: 1px solid #303338;
  border-radius: 1.5rem;
  padding: 0.5rem 0.8rem;
  background-color: transparent;
}

.c4 input {
  color: #fff;
  font-family: 'Inter',-apple-system,BlinkMacSystemFont,'Segoe UI','Roboto','Oxygen','Ubuntu','Cantarell','Fira Sans','Droid Sans','Helvetica Neue',sans-serif;
  font-size: 14px;
  font-weight: 500;
  text-align: left;
}

.c4 input::-webkit-input-placeholder {
  color: #A1A6B0;
}

.c4 input::-moz-placeholder {
  color: #A1A6B0;
}

.c4 input:-ms-input-placeholder {
  color: #A1A6B0;
}

.c4 input::placeholder {
  color: #A1A6B0;
}

.c4:hover {
  outline: 1px solid #fff;
}

.c4:focus {
  outline: 2px solid #fff;
}

.c5 {
  margin-right: 0.3rem;
}

.c9 {
  margin-left: 0.3rem;
}

.c7 {
  border-radius: 50%;
  height: 24px;
  width: 24px;
}

.c0 {
  box-sizing: 'border-box';
  min-width: 0;
  margin: 0;
  padding: 0;
}

.c10 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin: 0.75rem 0;
  width: 100%;
}

.c2 {
  margin-bottom: 0.75rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  color: #A1A6B0;
  font-size: 14px;
  font-weight: 500;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
}

.c11 {
  resize: none;
  box-sizing: border-box;
  border: 1px solid #303338;
  border-radius: 1.5rem;
  padding: 0.5rem 1.2rem;
  height: 7rem;
  width: 100%;
  background-color: transparent;
  color: #fff;
  font-family: 'Inter',Menlo,Monaco,Consolas,'Courier New',monospace;
  font-size: 14px;
  font-weight: 500;
}

<div>
   
  <div>
    <form>
      <div
        class="c0 c1"
      >
        <div
          class="c0 c2"
        >
          toAddress
        </div>
        <div
          class="c0 c3"
        >
          <div
            class="c4"
          >
            <div
              class="c5"
            >
              <div>
                <img
                  alt="Avatar"
                  class="c6 c7"
                  data-testid="avatar"
                  size="24"
                  src="https://metadata.ens.domains/mainnet/avatar/wagmi.eth"
                />
              </div>
            </div>
            <input
              class="c8"
              data-testid="input-to"
              name="to"
              placeholder="ethereumAddress"
              value="0x0000000000000000000000000000000000000000"
            />
            <div
              class="c9"
            >
              <div
                aria-label="clear address"
                class="c0 c10"
              >
                <svg
                  fill="none"
                  height="18"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="18"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <line
                    x1="18"
                    x2="6"
                    y1="6"
                    y2="18"
                  />
                  <line
                    x1="6"
                    x2="18"
                    y1="6"
                    y2="18"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="c0 c1"
      >
        <div
          class="c0 c2"
        >
          value (in wei)
        </div>
        <div
          class="c0 c3"
        >
          <div
            class="c4"
          >
            <div
              class="c5"
            />
            <input
              autocomplete="off"
              autocorrect="off"
              class="c8"
              data-testid="input-value"
              inputmode="decimal"
              maxlength="79"
              minlength="1"
              name="value"
              pattern="^[0-9]*[.,]?[0-9]*$"
              placeholder="0.0"
              spellcheck="false"
              type="text"
              value="1000000000000000000"
            />
            <div
              class="c9"
            />
          </div>
        </div>
      </div>
      <div
        class="c0 c1"
      >
        <div
          class="c0 c2"
        >
          data (hex)
        </div>
        <div
          class="c0 c3"
        >
          <textarea
            class="c11"
            data-testid="input-data"
            name="data"
            placeholder="data"
            spellcheck="false"
          >
            0x01
          </textarea>
        </div>
      </div>
      <button
        class="c12"
        data-testid="submit-rawtransaction"
        type="submit"
      >
        saveAction
      </button>
    </form>
  </div>
   
</div>
`;
